addons/binding/org.openhab.binding.unifi/src/main/java/org/openhab/binding/unifi/handler/UniFiControllerHandler.java

```diff
@@ -0,0 +1,172 @@
+            controller = null;
+            return;
+        }
+
+        scheduleRefreshJob();
+    }
+
+    @Override
+    public void dispose() {
+        cancelRefreshJob();
+    }
+
+    private void scheduleRefreshJob() {
+        logger.debug("Scheduling refresh job every {}s", config.getRefresh().longValue());
+        refreshJob = scheduler.scheduleWithFixedDelay(this, 0, config.getRefresh().longValue(), TimeUnit.SECONDS);
+    }
+
+    private void cancelRefreshJob() {
+        if (refreshJob != null) {
+            logger.debug("Cancelling refresh job");
+            refreshJob.cancel(true);
+            refreshJob = null;
+        }
+    }
+
+    private void refresh() {
+        ThingStatus status = getThing().getStatus();
+        ThingStatusDetail statusDetail = getThing().getStatusInfo().getStatusDetail();
+        String statusDescription = getThing().getStatusInfo().getDescription();
+
+        if (controller == null) {
+            // mgb: this should never happen
+            logger.warn("UniFi Controller is NULL - this should never happen!");
+        } else {
+            // mgb: we're offline, we need to try to login
+            if (status == INITIALIZING || (status == OFFLINE && statusDetail == COMMUNICATION_ERROR)) {
+                logger.debug("Trying to establish communication with the UniFi controller");
+                if (controller.login()) {
+                    status = ONLINE;
+                    statusDetail = ThingStatusDetail.NONE;
+                    statusDescription = null;
+                }
+            }
+
+            // mgb: only refresh if we're online
+            if (status == ONLINE) {
+                logger.debug("Refreshing the UniFi Controller {}", getThing().getUID());
+
+                try {
+                    // mgb: refresh the controller...
+                    controller.refresh();
+
+                    // mgb: ... then refresh all the clients
+                    for (Thing thing : getThing().getThings()) {
+                        ThingHandler handler = thing.getHandler();
+                        if (handler instanceof UniFiClientHandler) {
+                            UniFiClientHandler clientHandler = (UniFiClientHandler) handler;
+                            clientHandler.refresh();
+                        }
+                    }
+                } catch (UniFiException e) {
+                    logger.warn("Error refreshing the UniFi Controller {} - {}", getThing().getUID(), e.getMessage());
+                    status = OFFLINE;
+                    statusDetail = COMMUNICATION_ERROR;
+                    statusDescription = e.getMessage();
+                }
+            }
+        }
+
+        // mgb: update the status if it's changed
+        if (status != getThing().getStatus()) {
+            updateStatus(status, statusDetail, statusDescription);
+        }
+    }
+
+    @Override
+    public void run() {
```

Comment:
@wborn what is in your opinion the best thing to do here, implement runnable, just call this method, or directly call refresh and trust the logging from the framework and accept the fact that is might break the update if some uncatched exception occurs

Reply:
While some users were testing the binding, we had some NPEs being thrown (because of some weird data returned by the UniFi controller) that weren't being logged and the binding would just hang. After I figured out the subsequent calls to refresh were being suppressed because of the uncaught exceptions, I added the try/catch logic here.

I'm guessing things have changed with the logging, but how do you guys recommend I go about gracefully handling uncaught exceptions from a "refreshJob" point of view? If I do what you're suggesting, any error basically puts the binding in a frozen state and there will be no more updates until either the binding is reconfigured or OH is restarted.

Reply:
> that weren't being logged

Reply:
That makes sense.

What should I do about the refresh job? If I do it the way you suggest, any unhandled error will cause the binding to freeze.

Reply:
After some testing it seems that the `WrappedScheduledExecutorService` only logs the exception and after that further scheduled jobs no longer execute. :-(

I've created a simple exception binding with a thing handler that [schedules a job](https://github.com/wborn/openhab2-addons/blob/f2291e98a3d3832b6a2a4affa6903475ffb316df/addons/binding/org.openhab.binding.exception/src/main/java/org/openhab/binding/exception/handler/ExceptionThingHandler.java#L56) to throw a certain exception type. After the first execution it logs the error but thereafter the scheduled job no longer runs:

```
21:45:26.215 [WARN ] [ommon.WrappedScheduledExecutorService] - Scheduled runnable ended with an exception
java.lang.NullPointerException: NullPointerException Message
	at org.openhab.binding.exception.handler.ExceptionThingHandler.getConfiguredException(ExceptionThingHandler.java:93) ~[?:?]
	at org.openhab.binding.exception.handler.ExceptionThingHandler.runScheduledCommand(ExceptionThingHandler.java:72) ~[?:?]
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) [?:?]
	at java.util.concurrent.FutureTask.runAndReset(FutureTask.java:308) [?:?]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$301(ScheduledThreadPoolExecutor.java:180) [?:?]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:294) [?:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:?]
	at java.lang.Thread.run(Thread.java:748) [?:?]
```

So for now I agree you can better implement the try/catch yourself. 

@Hilbrand is this a bug in the WrappedScheduledExecutorService? If this is not a bug then I think the JavaDoc of that class should be updated so it is more clear what kind of burden is exactly removed for the user.